src/lib/profiles/data-management/Current/LoggingManagement.cpp

```diff
@@ -2146,6 +2246,171 @@ void CircularEventReader::Init(CircularEventBuffer * inBuf)
     }
 }
 
+WEAVE_ERROR CircularEventBuffer::SerializeEvents(TLVWriter & writer)
+{
+    WEAVE_ERROR err = WEAVE_NO_ERROR;
+
+    TLVType container;
+
+    uint8_t *inBufStart = mBuffer.QueueHead();
+    size_t inBufLen = mBuffer.DataLength();
+    uint8_t *bufStart = mBuffer.GetQueue();
+    size_t bufLen = mBuffer.GetQueueSize();
+    uint8_t *eventData = new uint8_t[inBufLen];
+
+    size_t initLen = inBufLen;
+    size_t remainingLen = 0;
+    if (initLen > bufStart + bufLen - inBufStart)
+    {
+        initLen = bufStart + bufLen - inBufStart;
+    }
+    if (inBufStart + inBufLen  - bufLen - bufStart > 0)
+    {
+        remainingLen = inBufStart + inBufLen  - bufLen - bufStart;
+    }
+
+    memcpy(eventData, inBufStart, initLen);
+    memcpy(eventData + initLen, bufStart, remainingLen);
+
+    err = writer.StartContainer(AnonymousTag, kTLVType_Structure, container);
+    SuccessOrExit(err);
+
+    err = writer.Put(ContextTag(kTag_PersistEvent_ImportanceLevel), (uint8_t)mImportance);
+    SuccessOrExit(err);
+
+    err = writer.PutBytes(ContextTag(kTag_PersistEvent_EventData), eventData, mBuffer.DataLength());
```

Comment:
Shouldn't bufStart be sufficient in putting bytes into the writer. Is eventData required with the associated memcpy?

Reply:
do you mean I should put two blocks of data under two different labels? I guess if we want to put two blocks under the same label, we need to do memcpy first, is that right

Reply:
We could eliminate the need to malloc by adding methods to WeaveTLVWriter for writing a byte string in multiple chunks.  E.g.:

TLVWriter::StartPutBytes(uint64_t tag, uint32_t totalLen);
TLVWriter::ContinuePutBytes(const uint8_t *buf, uint32_t len);

With this, you can put the event data into the writer directly from the queue buffer.